events.go

```diff
@@ -0,0 +1,196 @@
+	u, err := soap.ParseURL(vcenterURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing URL")
+	}
+
+	u.User = url.UserPassword(user, pass)
+	return govmomi.NewClient(ctx, u, insecure)
+}
+
+// eventStream is the main logic, blocking to receive and handle events from vCenter
+func eventStream(ctx context.Context, c *vim25.Client, controller *types.Controller) error {
+	// create event manager to consume events from vCenter
+	m := event.NewManager(c)
+
+	// get events for all types (i.e. RootFolder)
+	managedTypes := []vtypes.ManagedObjectReference{c.ServiceContent.RootFolder}
+	eventsPerPage := int32(1)
+	tail := true
+	force := true
+
+	recv := makeRecv(controller, m)
+	err := m.Events(ctx, managedTypes, eventsPerPage, tail, force, recv)
+	if err != nil {
+		return errors.Wrap(err, "error connecting to event-stream")
+	}
+	return nil
+}
+
+// makeRecv returns a event handler function called by the event manager on each event
+func makeRecv(controller *types.Controller, m *event.Manager) func(managedObjectReference vtypes.ManagedObjectReference, baseEvent []vtypes.BaseEvent) error {
+	return func(managedObjectReference vtypes.ManagedObjectReference, baseEvent []vtypes.BaseEvent) error {
+		log.Printf("Object %v", managedObjectReference)
+
+		for i, event := range baseEvent {
+			log.Printf("Event [%d] %v", i, event)
+
+			topic, message, err := handleEvent(event, m)
+			if err != nil {
+				log.Printf("error handling event: %s", err.Error())
+				continue
+			}
+			binaryMsg := []byte(message)
+			log.Printf("Message on topic: %s", topic)
+			controller.Invoke(topic, &binaryMsg)
+
+		}
+		return nil
+	}
+}
+
+func handleEvent(event vtypes.BaseEvent, m *event.Manager) (string, string, error) {
+	// Sanity check to avoid nil pointer exception
+	if event == nil {
+		return "", "", errors.New("event must not be nil")
+	}
+
+	// This pointer to the ManagedObjectReference will be used during the BaseEvent switch below
+	// If we don't find a MoRef in the event, *ref will be nil and not marshaled in the OutboundEvent
+	// Details on the ManagedObjectReference:
+	// https://code.vmware.com/docs/7371/vmware-vsphere-web-services-sdk-programming-guide-6-7-update-1#/doc/GUID-C9E81F17-2516-49EE-914F-EE9904258ED3.html
+	var ref *vtypes.ManagedObjectReference
+
+	// Get the type of the event, e.g. "VmPoweredOnEvent" which we'll use for subscribed topic matching
+	eventType := reflect.TypeOf(event).Elem().Name()
+	topic := convertToTopic(eventType)
+
+	// Retrieve user name and category from the event
+	user := event.GetEvent().UserName
+	createdTime := event.GetEvent().CreatedTime
+	category, err := m.EventCategory(context.Background(), event)
+	if err != nil {
+		return "", "", errors.Wrap(err, "error retrieving event category")
+	}
+
+	// Get the underlying concrete base event, e.g. VmEvent
+	// vSphere Web Service API Reference 6.7
+	// https://code.vmware.com/apis/358/vsphere#/doc/vim.event.Event.html
+	switch baseEvent := event.(type) {
```

Comment:
This needs to be extracted to a new function and unit tested. Single Responsibility Principle.